File: Advanced_Business_Development_with.NET/ChallengeMuttuApi/ChallengeMuttuApi/Model/Rastreamento.cs
// Path: ChallengeMuttuApi/Model/Rastreamento.cs - Este arquivo deve estar na pasta 'Model'
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ChallengeMuttuApi.Model
{
    /// <summary>
    /// Representa a entidade Rastreamento no banco de dados, mapeando para a tabela "TB_RASTREAMENTO".
    /// Contém dados de posicionamento interno (IPS) e GPRS (Latitude, Longitude, Altitude).
    /// </summary>
    [Table("TB_RASTREAMENTO")]
    public class Rastreamento
    {
        /// <summary>
        /// Construtor padrão da classe Rastreamento.
        /// </summary>
        public Rastreamento() { }

        /// <summary>
        /// Obtém ou define o identificador único do Rastreamento.
        /// Mapeia para a coluna "ID_RASTREAMENTO" (Chave Primária, gerada por identidade).
        /// </summary>
        [Key]
        [Column("ID_RASTREAMENTO")]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int IdRastreamento { get; set; }

        /// <summary>
        /// Obtém ou define a coordenada X do posicionamento interno (IPS).
        /// Mapeia para a coluna "IPS_X" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Coordenada X do posicionamento interno (IPS) com 4 inteiros e 3 decimais.'
        /// </summary>
        [Column("IPS_X", TypeName = "NUMBER(38,8)")] // Especifica o tipo para mapeamento decimal em Oracle
        [Required(ErrorMessage = "A coordenada IPS_X é obrigatória.")]
        public decimal IpsX { get; set; }

        /// <summary>
        /// Obtém ou define a coordenada Y do posicionamento interno (IPS).
        /// Mapeia para a coluna "IPS_Y" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Coordenada Y do posicionamento interno (IPS) com 4 inteiros e 3 decimais.'
        /// </summary>
        [Column("IPS_Y", TypeName = "NUMBER(38,8)")]
        [Required(ErrorMessage = "A coordenada IPS_Y é obrigatória.")]
        public decimal IpsY { get; set; }

        /// <summary>
        /// Obtém ou define a coordenada Z do posicionamento interno (IPS).
        /// Mapeia para a coluna "IPS_Z" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Coordenada Z do posicionamento interno (IPS) com 4 inteiros e 3 decimais.'
        /// </summary>
        [Column("IPS_Z", TypeName = "NUMBER(38,8)")]
        [Required(ErrorMessage = "A coordenada IPS_Z é obrigatória.")]
        public decimal IpsZ { get; set; }

        /// <summary>
        /// Obtém ou define a latitude do posicionamento GPRS.
        /// Mapeia para a coluna "GPRS_LATITUDE" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Latitude do posicionamento GPRS (ex: -90.000000 a +90.000000) com 6 decimais.'
        /// </summary>
        [Column("GPRS_LATITUDE", TypeName = "NUMBER(38,8)")]
        [Required(ErrorMessage = "A Latitude GPRS é obrigatória.")]
        public decimal GprsLatitude { get; set; }

        /// <summary>
        /// Obtém ou define a longitude do posicionamento GPRS.
        /// Mapeia para a coluna "GPRS_LONGITUDE" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Longitude do posicionamento GPRS (ex: -180.000000 a +180.000000) com 6 decimais.'
        /// </summary>
        [Column("GPRS_LONGITUDE", TypeName = "NUMBER(38,8)")]
        [Required(ErrorMessage = "A Longitude GPRS é obrigatória.")]
        public decimal GprsLongitude { get; set; }

        /// <summary>
        /// Obtém ou define a altitude do posicionamento GPRS.
        /// Mapeia para a coluna "GPRS_ALTITUDE" (NUMBER(38,8), Obrigatório).
        /// Comentário no DDL: 'Altitude do posicionamento GPRS (ex: 99999.99m) com 2 decimais.'
        /// </summary>
        [Column("GPRS_ALTITUDE", TypeName = "NUMBER(38,8)")]
        [Required(ErrorMessage = "A Altitude GPRS é obrigatória.")]
        public decimal GprsAltitude { get; set; }

        // Propriedades de Navegação (para Entity Framework Core)

        /// <summary>
        /// Coleção de entidades VeiculoRastreamento, representando o relacionamento muitos-para-muitos com Veículos.
        /// </summary>
        public ICollection<VeiculoRastreamento>? VeiculoRastreamentos { get; set; }
    }
}

File: Java_Advanced/motu-mvc-web/mottu-gradle/src/main/java/br/com/fiap/mottu/model/Rastreamento.java
package br.com.fiap.mottu.model;

import jakarta.persistence.*; // Importações do JPA para anotações de entidade
import lombok.*; // Importações do Lombok para reduzir boilerplate code (getters, setters, etc.)
import org.hibernate.annotations.CreationTimestamp; // Para gerar timestamp automaticamente na criação

import java.math.BigDecimal; // Para representar números decimais com precisão
import java.time.LocalDateTime; // Para representar data e hora sem fuso horário
import java.util.HashSet; // Implementação de Set
import java.util.Set; // Interface para coleções de elementos únicos

@Entity // Marca esta classe como uma entidade JPA (representa uma tabela no banco)
@Table(name = "TB_RASTREAMENTO") // Mapeia para a tabela TB_RASTREAMENTO
@Getter // Lombok: Gera automaticamente os métodos getters
@Setter // Lombok: Gera automaticamente os métodos setters
@NoArgsConstructor // Lombok: Gera um construtor sem argumentos
@AllArgsConstructor // Lombok: Gera um construtor com todos os argumentos
@Builder // Lombok: Habilita o padrão de projeto Builder para facilitar a criação de instâncias
@ToString(exclude = {"veiculoRastreamentos"}) // Lombok: Gera o método toString, excluindo o campo veiculoRastreamentos para evitar recursão infinita
@EqualsAndHashCode(onlyExplicitlyIncluded = true) // Lombok: Gera equals e hashCode baseados apenas nos campos anotados com @EqualsAndHashCode.Include
public class Rastreamento {

    @Id // Marca o campo idRastreamento como a chave primária da tabela
    @GeneratedValue(strategy = GenerationType.IDENTITY) // Configura a geração automática do ID pelo banco (identidade)
    @Column(name = "ID_RASTREAMENTO") // Mapeia para a coluna ID_RASTREAMENTO
    @EqualsAndHashCode.Include // Inclui este campo no cálculo de equals e hashCode
    private Long idRastreamento; // Identificador único do rastreamento

    // Precisão (total de dígitos) e escala (dígitos após a vírgula) ajustadas conforme DDL da tabela.
    // Se o DDL usar NUMBER(38,8), as precisões e escalas aqui deveriam ser maiores.
    // Atualmente, está configurado para refletir as classes Java originais (ex: 7,3 para IPS).
    @Column(name = "IPS_X", nullable = false, precision = 7, scale = 3) // Coordenada X do Indoor Positioning System
    private BigDecimal ipsX;

    @Column(name = "IPS_Y", nullable = false, precision = 7, scale = 3) // Coordenada Y do Indoor Positioning System
    private BigDecimal ipsY;

    @Column(name = "IPS_Z", nullable = false, precision = 7, scale = 3) // Coordenada Z do Indoor Positioning System
    private BigDecimal ipsZ;
    @Column(name = "GPRS_LATITUDE", nullable = false, precision = 11, scale = 6) // Latitude obtida via GPRS
    private BigDecimal gprsLatitude;

    @Column(name = "GPRS_LONGITUDE", nullable = false, precision = 11, scale = 6) // Longitude obtida via GPRS
    private BigDecimal gprsLongitude;

    @Column(name = "GPRS_ALTITUDE", nullable = false, precision = 7, scale = 2) // Altitude obtida via GPRS
    private BigDecimal gprsAltitude;

    @CreationTimestamp // Hibernate: Preenche automaticamente com o timestamp no momento da criação da entidade
    @Column(name = "DATA_HORA_REGISTRO", nullable = false, updatable = false) // Data e hora do registro do rastreamento
    private LocalDateTime dataHoraRegistro; // NOVO CAMPO

    // Relacionamento bidirecional com VeiculoRastreamento (tabela de junção)
    @OneToMany(mappedBy = "rastreamento", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default // Lombok: Inicializa o Set para evitar NullPointerExceptions
    private Set<br.com.fiap.mottu.model.relacionamento.VeiculoRastreamento> veiculoRastreamentos = new HashSet<>();
}